### PR TITLE
Structure Generation Fixes

### PR DESCRIPTION
**Pull request by Dariensg**
-Made Strider Altars check for air above the lava so that they generate in the open and not in the wall
-Created new Generation Validity checks for all other structures that look for a fairly flat and open area of land to spawn in with enough vertical space for the structure
-Changed Structure spacing and separation to match the new generation checks to get reasonable numbers of structures
-Removed x and z coordinate randomization of structures as it was unnecessary
-Had several headaches but finally got it working

Lots of credit goes to TelepathicGrunt for all the help with this one

### CODE REVIEW

File: src/main/java/com/nekomaster1000/infernalexp/world/gen/structures/BastionOutpostStructure.java [hunk 1]
```diff
@@ -35,32 +43,114 @@ public BastionOutpostStructure(Codec<NoFeatureConfig> codec) {
...
+            int posZ = chunkZ << 4;
+            int posY = getYPos(chunkGenerator, posX, posZ, random);
+
+            // Checks 9 points within about a chunk of the initial location
+            for (int curX = posX - 8; curX <= posX + 8; curX += 8) {
+                for (int curZ = posZ - 8; curZ <= posZ + 8; curZ += 8) {
+
+                    // Starts 5 blocks below to check for solid land in each column
+                    BlockPos.Mutable mutable = new BlockPos.Mutable();
+                    mutable.setPos(curX, posY - 5, curZ);
```
Comment (SwanX1) [inline src/main/java/com/nekomaster1000/infernalexp/world/gen/structures/BastionOutpostStructure.java]:
```suggestion
                    BlockPos.Mutable mutable = new BlockPos.Mutable(curX, posY - 5, curZ);
```

File: src/main/java/com/nekomaster1000/infernalexp/world/gen/structures/BastionOutpostStructure.java [hunk 2]
```diff
@@ -35,32 +43,114 @@ public BastionOutpostStructure(Codec<NoFeatureConfig> codec) {
...
+                }
+            }
+        }
+        else {
+            return false;
+        }
+
+        return true;
+    }
+    public int getYPos(ChunkGenerator chunkGenerator, int x, int z, SharedSeedRandom random) {
```
Comment (SwanX1) [inline src/main/java/com/nekomaster1000/infernalexp/world/gen/structures/BastionOutpostStructure.java]:
```suggestion
    }
    
    public int getYPos(ChunkGenerator chunkGenerator, int x, int z, SharedSeedRandom random) {
```

Comment (SwanX1) [reply src/main/java/com/nekomaster1000/infernalexp/world/gen/structures/BastionOutpostStructure.java]: Documenting this method would be very helpful for future contributors.

Comment (SwanX1) [reply src/main/java/com/nekomaster1000/infernalexp/world/gen/structures/BastionOutpostStructure.java]: Shouldn't this method be static?

File: src/main/java/com/nekomaster1000/infernalexp/world/gen/structures/BastionOutpostStructure.java [hunk 3]
```diff
@@ -35,32 +43,114 @@ public BastionOutpostStructure(Codec<NoFeatureConfig> codec) {
...
+                        BlockState state = blockView.getBlockState(mutable);
+                        if (!state.isAir()) {
+                            return false;
+                        }
+                        mutable.move(Direction.UP);
+                    }
+                }
+            }
+        }
+        else {
```
Comment (SwanX1) [inline src/main/java/com/nekomaster1000/infernalexp/world/gen/structures/BastionOutpostStructure.java]:
```suggestion
        } else {
```

**Review by SwanX1 — CHANGES_REQUESTED** [review]
the same comments from `BastionOutpostStructure::func_230363_a_` and `BastionOutpostStructure::getYPos` apply to `GlowstoneCanyonRuinStructure`, `SoulSandValleyRuinStructure`, and `StriderAltarStructure`.

I'd also recommend refactoring the code to make all the previously mentioned classes inherit a class that has their shared code, like `func_230363_a_`, `getYPos`, and the inner class `Start`.
The inheritance would be as such
 - `SpecificStructure`
   - `SolidGroundSpawningStructure<Config>` <- new class
     - `IEStructure<Config>`

File: src/main/java/com/nekomaster1000/infernalexp/world/gen/structures/BastionOutpostStructure.java (hunk 1 above)
Comment (Hellion-Games) [reply src/main/java/com/nekomaster1000/infernalexp/world/gen/structures/BastionOutpostStructure.java]: The thing about moving func_240363_a_ to a super class is that each structure type uses different values for the checks, so I'd still just have to override it in each subclass specific structure anyway. I could move the getYPos method, but that's about the only one that makes sense.

Comment (SwanX1) [reply src/main/java/com/nekomaster1000/infernalexp/world/gen/structures/BastionOutpostStructure.java]: Couldn't you just add new arguments for those values?

Comment (Hellion-Games) [reply src/main/java/com/nekomaster1000/infernalexp/world/gen/structures/BastionOutpostStructure.java]: Fair enough, but there's a lot of them. There's how far away to go from the initial point, how much to increment after each sample, how vertically to check for land, and how vertically to check for air above the structure. That feels like a bit much but I can do it if you think I should.

Comment (SwanX1) [reply src/main/java/com/nekomaster1000/infernalexp/world/gen/structures/BastionOutpostStructure.java]: Hmm... if it's that much, then don't implement it in the superclass. (also this thread is in the wrong place, this is about a suggestion for the code on line 70 LMAO)

**Comment by SwanX1** [comment]
Please fix the compilation issues

**Comment by Dariensg** [comment]
> 
> 
> Please fix the compilation issues

Weird. It launches fine for me from my dev environment and the error that the build failed on was from a mixin with ItemModelsProperties it looked like. I'm trying re-running it.

**Comment by SwanX1** [comment]
Merged in commit d18021f9b0b5f7733d4536c98bed09dfb458473f, pull request doesn't show as merged due to me removing the last two merge commits to avoid unnecessary commit spam.

**Pull request closed without merging**